components/speedreader/rust/ffi/src/speedreader.cc

```diff
@@ -0,0 +1,147 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/speedreader/rust/ffi/include/speedreader.h"
+
+#include <iostream>
+#include "base/logging.h"
+
+#include "brave/components/speedreader/rust/ffi/include/speedreader_ffi.h"
+
+namespace speedreader {
+
+SpeedReader::SpeedReader() : raw(speedreader_new()) {}
+SpeedReader::SpeedReader(const char* whitelist_serialized,
+                         size_t whitelist_size)
+    : raw(with_whitelist(whitelist_serialized, whitelist_size)) {}
+
+bool SpeedReader::deserialize(const char* data, size_t data_size) {
+  auto* new_raw = with_whitelist(data, data_size);
+  if (new_raw != nullptr) {
+    speedreader_free(raw);
+    raw = new_raw;
+    return true;
+  } else {
+    VLOG(2) << __func__ << " deserialization failed: " << TakeLastError();
+    return false;
+  }
+}
+
+SpeedReader::~SpeedReader() {
+  speedreader_free(raw);
+}
+
+bool SpeedReader::ReadableURL(const std::string& url) {
+  return url_readable(raw, url.c_str(), url.length());
+}
+
+RewriterType SpeedReader::RewriterTypeForURL(const std::string& url) {
+  return find_type(raw, url.c_str(), url.length());
+}
+
+std::unique_ptr<Rewriter> SpeedReader::RewriterNew(const std::string& url) {
+  return std::make_unique<Rewriter>(raw, url, RewriterType::RewriterUnknown);
+}
+
+std::unique_ptr<Rewriter> SpeedReader::RewriterNew(const std::string& url,
+                                                   RewriterType rewriter_type) {
+  return std::make_unique<Rewriter>(raw, url, rewriter_type);
+}
+
+std::unique_ptr<Rewriter> SpeedReader::RewriterNew(
+    const std::string& url,
+    RewriterType rewriter_type,
+    void (*output_sink)(const char*, size_t, void*),
+    void* output_sink_user_data) {
+  return std::make_unique<Rewriter>(raw, url, rewriter_type, output_sink,
+                                    output_sink_user_data);
+}
+
+// static
+std::string SpeedReader::TakeLastError() {
+  auto* error = take_last_speedreader_error();
+  if (error) {
+    std::string err(error->data, error->len);
+    speedreader_str_free(*error);
+    return err;
+  } else {
+    return "";
+  }
+}
+
+Rewriter::Rewriter(C_SpeedReader* speedreader,
+                   const std::string& url,
+                   RewriterType rewriter_type)
+    : output_(""),
+      ended_(false),
```

Comment:
you can save a few lines here and there by initializing these in the class declaration